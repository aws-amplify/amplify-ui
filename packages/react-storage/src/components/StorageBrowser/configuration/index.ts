export { createConfigurationProvider } from './createConfigurationProvider';
export { useGetActionInput } from './context';
export { GetActionInput } from './types';
