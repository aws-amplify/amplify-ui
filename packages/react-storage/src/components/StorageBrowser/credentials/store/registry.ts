import {
  AWSTemporaryCredentials,
  StorageValidationErrorCode,
  assertValidationError,
} from '../../storage-internal';
import { GetLocationCredentials, CredentialsLocation } from '../types';

import {
  LruLocationCredentialsStore,
  fetchNewValue,
  getCacheValue,
  initStore,
} from './store';

interface StoreRegistrySymbol {
  readonly value: symbol;
}

/**
 * Keep all cache records for all instances of credentials store in a singleton
 * so we can reliably de-reference from the memory when we destroy a store
 * instance.
 */
const storeRegistry = new WeakMap<
  StoreRegistrySymbol,
  LruLocationCredentialsStore
>();

/**
 * @internal
 */
export const createStore = (
  refreshHandler: GetLocationCredentials,
  size?: number
): StoreRegistrySymbol => {
  const storeSymbol = { value: Symbol('LocationCredentialsStore') };
  storeRegistry.set(storeSymbol, initStore(refreshHandler, size));

  return storeSymbol;
};

const getCredentialsStore = (storeSymbol: StoreRegistrySymbol) => {
  assertValidationError(
    storeRegistry.has(storeSymbol),
    StorageValidationErrorCode.LocationCredentialsStoreDestroyed
  );

  return storeRegistry.get(storeSymbol)!;
};

/**
 * @internal
 */
export const getValue = async (input: {
  storeSymbol: StoreRegistrySymbol;
  location: CredentialsLocation;
  forceRefresh: boolean;
}): Promise<{ credentials: AWSTemporaryCredentials }> => {
  const { storeSymbol: storeReference, location, forceRefresh } = input;
  const store = getCredentialsStore(storeReference);
  if (!forceRefresh) {
    const credentials = getCacheValue(store, location);
      if (credentials !== null) {
        return { credentials };
      }
  }

  return fetchNewValue(store, location);
};

export const removeStore = (storeSymbol: StoreRegistrySymbol): void => {
  storeRegistry.delete(storeSymbol);
};
