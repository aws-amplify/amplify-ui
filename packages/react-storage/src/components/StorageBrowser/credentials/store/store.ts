import { LocationPermissions } from '../../credentials/types';
import {
  AWSTemporaryCredentials,
  StorageValidationErrorCode,
  assertValidationError,
} from '../../storage-internal';
import { GetLocationCredentials } from '../types';

import {
  CREDENTIALS_REFRESH_WINDOW_MS,
  CREDENTIALS_STORE_DEFAULT_SIZE,
} from './constants';

interface CredentialsLocation {
  scope: string;
  permissions: LocationPermissions;
}

interface StoreValue extends CredentialsLocation {
  credentials?: AWSTemporaryCredentials;
  inflightCredentials?: Promise<{ credentials: AWSTemporaryCredentials }>;
}

type S3Uri = string;

type SerializedPermissions = string;

type CacheKey = `${S3Uri}_${SerializedPermissions}`;

const serializedPermissions = (
  permissions: LocationPermissions
): SerializedPermissions => permissions.sort().join('_');

const createCacheKey = (location: CredentialsLocation): CacheKey =>
  `${location.scope}_${serializedPermissions(location.permissions)}`;

/**
 * LRU implementation for Location Credentials Store
 * O(n) for get and set for simplicity.
 *
 * @internal
 */
export interface LruLocationCredentialsStore {
  capacity: number;
  refreshHandler: GetLocationCredentials;
  values: Map<CacheKey, StoreValue>;
}

const pastTTL = (credentials: AWSTemporaryCredentials) => {
  const { expiration } = credentials;

  return expiration.getTime() - CREDENTIALS_REFRESH_WINDOW_MS <= Date.now();
};

const setCacheRecord = (
  store: LruLocationCredentialsStore,
  key: CacheKey,
  value: StoreValue
): void => {
  if (store.capacity === store.values.size) {
    // Pop least used entry. The Map's key are in insertion order.
    // So first key is the last recently inserted.
    const [oldestKey] = store.values.keys();
    store.values.delete(oldestKey);
    // TODO(@AllanZhengYP): Add log info when record is evicted.
  }
  // Add latest used value to the cache.
  store.values.set(key, value);
};

const dispatchRefresh = (
  refreshHandler: GetLocationCredentials,
  value: StoreValue,
  onRefreshFailure: () => void
) => {
  if (value.inflightCredentials) {
    return value.inflightCredentials;
  }

  value.inflightCredentials = (async () => {
    try {
      const { credentials } = await refreshHandler({
        scope: value.scope,
        permissions: value.permissions,
      });
      value.credentials = credentials;

      return { credentials };
    } catch (e) {
      onRefreshFailure();
      throw e;
    } finally {
      value.inflightCredentials = undefined;
    }
  })();

  return value.inflightCredentials;
};

/**
 * @internal
 */
export const initStore = (
  refreshHandler: GetLocationCredentials,
  size = CREDENTIALS_STORE_DEFAULT_SIZE
): LruLocationCredentialsStore => {
  assertValidationError(
    size > 0,
    StorageValidationErrorCode.InvalidLocationCredentialsCacheSize
  );

  return {
    capacity: size,
    refreshHandler,
    values: new Map<CacheKey, StoreValue>(),
  };
};

export const getCacheValue = (
  store: LruLocationCredentialsStore,
  location: CredentialsLocation
): AWSTemporaryCredentials | null => {
  const cacheKey = createCacheKey(location);
  const cachedValue = store.values.get(cacheKey);
  const cachedCredentials = cachedValue?.credentials;
  if (!cachedCredentials) {
    return null;
  }

  // Delete and re-insert to key to map to indicate a latest reference in LRU.
  store.values.delete(cacheKey);
  if (!pastTTL(cachedCredentials)) {
    // TODO(@AllanZhengYP): If the credential is still valid but will expire
    // soon, we should return credentials AND dispatch a refresh.
    store.values.set(cacheKey, cachedValue);

    return cachedCredentials;
  }

  return null;
};

/**
 * Fetch new credentials value with refresh handler and cache the result in
 * LRU cache.
 * @internal
 */
export const fetchNewValue = async (
  store: LruLocationCredentialsStore,
  location: CredentialsLocation
): Promise<{ credentials: AWSTemporaryCredentials }> => {
  const storeValues = store.values;
  const key = createCacheKey(location);
  if (!storeValues.has(key)) {
    const newStoreValue: StoreValue = {
      scope: location.scope,
      permissions: location.permissions,
    };
    setCacheRecord(store, key, newStoreValue);
  }
  const storeValue = storeValues.get(key)!;

  return dispatchRefresh(store.refreshHandler, storeValue, () => {
    store.values.delete(key);
  });
};
