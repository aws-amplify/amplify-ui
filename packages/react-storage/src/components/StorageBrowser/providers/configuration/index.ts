export { useGetActionInput } from './context';
export { createConfigurationProvider } from './createConfigurationProvider';
export { CredentialsProviderProps, RegisterAuthListener } from './credentials';
export { GetActionInput } from './types';
