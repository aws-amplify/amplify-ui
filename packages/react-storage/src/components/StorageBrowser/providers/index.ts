export {
  createConfigurationProvider,
  CredentialsProviderProps,
  RegisterAuthListener,
} from './configuration';
export {
  FileItem,
  FileItems,
  StoreProviderProps,
  StoreProvider,
} from './store';
