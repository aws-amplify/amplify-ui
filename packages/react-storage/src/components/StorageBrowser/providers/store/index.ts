export { FileItem, FileItems } from './files';
export { StoreProvider, StoreProviderProps } from './StoreProvider';
export { useStore, UseStoreState } from './useStore';
