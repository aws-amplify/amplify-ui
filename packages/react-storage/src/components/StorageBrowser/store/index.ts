export { StoreProvider, useStore } from './context';
export {
  LocationEventValue,
  LocationState,
  LocationValue,
  StorageBrowserEventValue,
  StorageBrowserValue,
  StoreProviderProps,
  StoreState,
} from './types';
