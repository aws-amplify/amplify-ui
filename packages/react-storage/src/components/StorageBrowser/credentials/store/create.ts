import {
  LocationCredentialsProvider,
  StorageValidationErrorCode,
  assertValidationError,
} from '../../storage-internal';
import {
  CreateLocationCredentialsStoreInput,
  CredentialsLocation,
  LocationCredentialsStore,
} from '../types';

import { createStore, getValue, removeStore } from './registry';

const validateS3Uri = (uri: string): void => {
  const s3UrlSchemaRegex = /^s3:\/\/[^/]+/;
  assertValidationError(
    s3UrlSchemaRegex.test(uri),
    StorageValidationErrorCode.InvalidS3Uri
  );
};

export const createLocationCredentialsStore = (
  input: CreateLocationCredentialsStoreInput
): LocationCredentialsStore => {
  const storeSymbol = createStore(input.handler);

  const store = {
    getProvider(providerLocation: CredentialsLocation) {
      const locationCredentialsProvider = async ({
        forceRefresh = false,
      }: Parameters<LocationCredentialsProvider>[0] = {}) => {
        validateS3Uri(providerLocation.scope);

        // TODO(@AllanZhengYP): validate the action bucket and paths matches provider scope.
        return getValue({
          storeSymbol,
          location: { ...providerLocation },
          forceRefresh,
        });
      };

      return locationCredentialsProvider;
    },

    destroy() {
      removeStore(storeSymbol);
    },
  };

  return store;
};
