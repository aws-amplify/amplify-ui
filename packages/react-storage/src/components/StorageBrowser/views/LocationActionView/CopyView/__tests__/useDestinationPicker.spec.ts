import { act, renderHook, waitFor } from '@testing-library/react';

import * as AmplifyReactCore from '@aws-amplify/ui-react-core';

import * as Store from '../../../../providers/store';
import * as Config from '../../../../providers/configuration';
import {
  DEFAULT_LIST_OPTIONS,
  useDestinationPicker,
} from '../useDestinationPicker';

const mockDispatchStoreAction = jest.fn();
const mockHandleList = jest.fn();
const config = {
  accountId: '123456789012',
  bucket: 'bucket',
  credentials: jest.fn(),
  region: 'us-west-2',
};

jest.useFakeTimers();
jest.setSystemTime(1731366223230);

const mockItems = [
  {
    key: 'prefix1/',
    lastModified: new Date(),
    id: 'id',
    size: 10,
    type: 'FOLDER',
  },
  {
    key: 'prefix2/',
    lastModified: new Date(),
    id: 'id',
    size: 10,
    type: 'FOLDER',
  },
];

describe('useDestinationPicker', () => {
  beforeEach(() => {
    jest.clearAllMocks();

    jest.spyOn(Store, 'useStore').mockReturnValue([
      {
        actionType: 'COPY',
        files: [],
        location: {
          current: {
            prefix: 'test-prefix/',
            bucket: 'bucket',
            id: 'id',
            permission: 'READWRITE',
            type: 'PREFIX',
          },
          path: '',
          key: 'test-prefix/',
        },
        locationItems: {
          fileDataItems: [
            {
              key: 'prefixer/test-file.txt',
              fileKey: 'test-file.txt',
              lastModified: new Date(),
              id: 'id',
              size: 10,
              type: 'FILE',
            },
          ],
        },
      },
      mockDispatchStoreAction,
    ]);

    jest.spyOn(Config, 'useGetActionInput').mockReturnValue(() => config);
  });

  it('should return the correct initial state', async () => {
    jest.spyOn(AmplifyReactCore, 'useDataState').mockReturnValueOnce([
      {
        data: {
          items: mockItems,
          nextToken: 'token',
        },
        hasError: false,
        isLoading: false,
        message: undefined,
      },
      mockHandleList,
    ]);

    const { result } = renderHook(() =>
      useDestinationPicker({
        destinationList: ['prefix1'],
      })
    );

    await waitFor(() => {
      expect(result.current).toMatchSnapshot();
    });
  });

  it('should call handleList once per destinationList change', async () => {
    jest.spyOn(AmplifyReactCore, 'useDataState').mockReturnValue([
      {
        data: {
          items: mockItems,
          nextToken: 'token',
        },
        hasError: false,
        isLoading: false,
        message: undefined,
      },
      mockHandleList,
    ]);

    const { rerender } = renderHook(
      (
        props: { destinationList: string[] } = {
          destinationList: ['prefix1'],
        }
      ) => useDestinationPicker(props)
    );
    await waitFor(() => {
      rerender({
        destinationList: ['prefix1', 'subfolder1'],
      });
      rerender({
        destinationList: ['prefix1'],
      });
      rerender({
        destinationList: ['prefix1'],
      });
      expect(mockHandleList).toHaveBeenCalledTimes(3);
    });
  });

  it('should handle search', () => {
    jest.spyOn(AmplifyReactCore, 'useDataState').mockReturnValueOnce([
      {
        data: {
          items: mockItems,
          nextToken: 'token',
        },
        hasError: false,
        isLoading: false,
        message: undefined,
      },
      mockHandleList,
    ]);
    const { result } = renderHook(() =>
      useDestinationPicker({
        destinationList: ['prefix1'],
      })
    );

    act(() => {
      result.current.onSearchQueryChange('moo');
    });

    act(() => {
      result.current.onSearch();
    });

    expect(mockHandleList).toHaveBeenCalledWith({
      config,
      options: {
        ...DEFAULT_LIST_OPTIONS,
        exclude: 'FILE',
        search: { filterKey: 'key', query: 'moo' },
      },
      prefix: 'prefix1/',
    });

    // clearing search refreshes the list
    act(() => {
      result.current.onSearchClear();
    });

    expect(mockHandleList).toHaveBeenCalledWith(
      expect.objectContaining({
        options: expect.objectContaining({
          refresh: true,
        }),
      })
    );
  });
});
