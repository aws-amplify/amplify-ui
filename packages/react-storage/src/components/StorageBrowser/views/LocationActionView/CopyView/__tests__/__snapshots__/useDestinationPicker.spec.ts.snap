// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useDestinationPicker should return the correct initial state 1`] = `
{
  "bucket": "bucket",
  "currentPage": 1,
  "hasError": false,
  "hasNextToken": true,
  "highestPageVisited": 1,
  "isLoading": false,
  "items": [
    {
      "id": "id",
      "key": "prefix1/",
      "lastModified": 2024-11-11T23:03:43.230Z,
      "size": 10,
      "type": "FOLDER",
    },
    {
      "id": "id",
      "key": "prefix2/",
      "lastModified": 2024-11-11T23:03:43.230Z,
      "size": 10,
      "type": "FOLDER",
    },
  ],
  "message": undefined,
  "onPaginate": [Function],
  "onSearch": [Function],
  "onSearchClear": [Function],
  "onSearchQueryChange": [Function],
  "pageItems": [
    {
      "id": "id",
      "key": "prefix1/",
      "lastModified": 2024-11-11T23:03:43.230Z,
      "size": 10,
      "type": "FOLDER",
    },
    {
      "id": "id",
      "key": "prefix2/",
      "lastModified": 2024-11-11T23:03:43.230Z,
      "size": 10,
      "type": "FOLDER",
    },
  ],
  "resetSearch": [Function],
  "searchQuery": "",
}
`;
